src/lib/nvme_pt_ioctl.c

```diff
@@ -727,6 +733,128 @@ static int validate_sessn(struct opal_device *dev)
 	return 0;
 }
 
+static int parse_tper_host_prop(struct sed_device *device)
+{
+	struct opal_device *dev = device->priv;
+	int payload_len = dev->payload.len;
+	struct sed_tper_properties *tper = &device->discv.sed_tper_props;
+
+	/* TPer properties are returned as key-value pairs */
+	for (int i = 0; i < payload_len; i++) {
+		if(resp_token_match(dev->payload.tokens[i], OPAL_STARTNAME)) {
+			int jmp = get_payload_string(dev, i + 1);
+			int key_len = dev->payload.tokens[i + 1]->len - jmp;
+			char *key = malloc(key_len);
```

Comment:
Current code does following for each name-value pair:
1. calculate key len
2. allocate buffer,
3. process property
4. free memory.

Instead of calculating key_len each time you could assume some max e.g. 64B and preallocate buffer, and reause buffer for each iteration(BTW I think that spec defines max len as 32B ). When you recognize discover property name larger than this buffer code could simply crash indicating that buffer size needs to be larger . This approach could help remove significant complexity from the loop of allocating and deallocating memory and handling errors related to it.